dts/bindings/fpga/lattice,ice40-fpga.yaml

```diff
@@ -0,0 +1,77 @@
+# Copyright (c) 2022 Friedt Professional Engineering Services, Inc
+# SPDX-License-Identifier: Apache-2.0
+
+description: Lattice iCE40 FPGA
+
+compatible: "lattice,ice40-fpga"
+
+include: spi-device.yaml
+
+properties:
+
+    cdone-gpios:
+      type: phandle-array
+      required: true
+      description: |
+        Configuration Done output from iCE40.
+        
+        Example usage: 
+        cdone-gpios = <&gpio0 0 GPIO_ACTIVE_HIGH>;
+
+    creset-gpios:
+      type: phandle-array
+      required: true
+      description: |
+        Configuration Reset input on iCE40.
+        
+        Example usage: 
+        creset-gpios = <&gpio0 1 GPIO_ACTIVE_LOW>;
+    
+    cs-as-gpio-state:
+      type: int
+      required: true
+      description: |
+        The iCE40 requires the SPI controller to expose an
+        alternate function where the chip-select pin is actually
+        GPIO output.
+
+        Set this to the array index of the appropriate pinctrl state
+        in the SPI controller.
+
+        Example usage:        
+        &spi2 {
+          ...
+          pinctrl-2 = <&spim2_cs_as_gpio>;
+          ...
+          fpga0: fpga@0 {
+            ...
+        		cs-as-gpio-state = <2>;
```

Comment:
there's no need to store state index, you can just `pinctrl_apply_state(pcfg, PINCTRL_STATE_CS_AS_GPIO)`.

Reply:
Would that be defined in the .dts?

Reply:
`pinctrl_apply_state` performs a runtime lookup.

Reply:
Is it OK that this type of pinctrl call is done outside of the spi controller and in a peripheral driver?

FWIR, it might require CONFIG_PINCTRL_DYNAMIC

Reply:
Hmm tbh pinctrl was always designed assuming only one driver controls the pins of a peripheral. Dynamic pinctrl is designed to route to a new set of pins at runtime, not  to allow applying states out of driver's control. I guess one option would be to extend the SPI driver with a custom call that _releases_ CS pin.

Reply:
Yeah, I'm not really sure what the best way forward here is in terms of pinctrl. This approach was also used for an i2c driver that required similar handshaking by @henrikbrixandersen. Some hardware is seemingly quite capable of using this approach.

In terms of maintenance, this type of situation is quite similar to when `arbitration` is required between two competing RF stacks or even competing transceivers on the same antenna; API consumers might believe the HW is doing what is expected while it might be doing but, in actuality, the output pads (or perhaps an external switch or PA) is disconnected or disabled.

Technically speaking a pseudo-device could be created that could (optionally) be backed by a GPIO. If there is no GPIO, then it's simply software arbitration. Driver DT could inherit this type of thing as a dts binding (phandle), and could return e.g. `-EBUSY` when they do not hold control. Generally, it's going to be platform or application dependent how this type of arbitration works.